### ISSUE TITLE
First-class support for lazy named tag evaluation?

### ISSUE THREAD

**Issue by eprothro** [rocketjob/semantic_logger#117]
If there's a mailing list or other avenue you prefer for this enhancement discussion, please let me know.

We have a need for lazy evaluation of named tags whose values are discovered during the execution of the block. We can work around this or patch for our specific needs, but I wanted to see if this has been discussion before or anyone else has this need.

A contrived example:

```ruby
require 'semantic_logger'
SemanticLogger.add_appender(io: $stdout, level: :debug, formatter: :json)
LOGGER = SemanticLogger['Contrived']
LOGGER.level = :debug

Thread.current[:execution_context] = {}

class Scope1
  def execute
    Thread.current[:execution_context][:discovered_tag] = "value"
  end
end

class Scope2
  def execute
    LOGGER.debug('some.event')
  end
end

LOGGER.tagged(Thread.current[:execution_context]) do
  Scope1.new.execute
  Scope2.new.execute
end
```

Clearly `some.event` will be logged without the `discovered_tag`, since we didn't know about it when calling `tagged`.

```json
{
	"host": "local",
	"application": "Semantic Logger",
	"timestamp": "2018-09-10T20:26:47.833371Z",
	"level": "debug",
	"level_index": 1,
	"pid": 23155,
	"thread": "70338952296820",
	"name": "Contrived",
	"message": "some.event"
}
```

I assume the recommended way to do this is using `push_named_tags` and ensuring they are later popped appropriately.

```ruby
require 'semantic_logger'
SemanticLogger.add_appender(io: $stdout, level: :debug, formatter: :json)
LOGGER = SemanticLogger['Contrived']
LOGGER.level = :debug
Thread.current[:contrived_extra_tags] = 0

class Scope1
  def execute
    SemanticLogger.push_named_tags({some_tag: "discovered"})
    Thread.current[:contrived_extra_tags] += 1
  end
end

class Scope2
  def execute
    LOGGER.debug('some.event')
  end
end

LOGGER.tagged do
  Scope1.new.execute
  Scope2.new.execute
end
SemanticLogger.pop_named_tags(Thread.current[:contrived_extra_tags])

```

Any thoughts regarding adding interface that would allow for lazy evaluation of certain named tags, and/or the ability to add a tag that only lives for the scope of the current execution (tagged block)?

**Comment by reidmorrison** [comment]
Not sure I understand the requirement for lazy evaluation of named tags. If you want named tags within one thread to be automatically copied to a spawned thread within that same block, then checkout: http://reidmorrison.github.io/parallel_minion/

Parallel Minion was created as a thin wrapper on top of Thread. It adds logging around thread calls and copies logging named tags.

**Comment by eprothro** [comment]
@reidmorrison thanks for the parallel minion info. 

Since creation of this PR I've realized that what we were searching for architecturally was not having to use the closure style of `tagged`. Hopefully what I mean by that is clear in the above example(s).

Once we recognized that, we realized the implications were not ideal and we prefer the simplicity / safety of the closure. We've rearchitected our implementation to work with the `tagged` architecture.
